File: TextosJapones/src/main/java/org/jisho/textosJapones/fileparse/TarParse.java
package org.jisho.textosJapones.fileparse;

import org.apache.commons.compress.archivers.tar.TarArchiveEntry;
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream;
import org.jisho.textosJapones.util.NaturalOrderComparator;
import org.jisho.textosJapones.util.Util;

import java.io.*;
import java.util.*;

public class TarParse implements Parse {

	private List<TarEntry> mEntradas;
	private List<TarEntry> mLegendas;

	private class TarEntry {
		final TarArchiveEntry entry;
		final byte[] bytes;

		public TarEntry(TarArchiveEntry entry, byte[] bytes) {
			this.entry = entry;
			this.bytes = bytes;
		}
	}

	@Override
	public void parse(File file) throws IOException {
		mEntradas = new ArrayList<>();

		BufferedInputStream fis = new BufferedInputStream(new FileInputStream(file));
		TarArchiveInputStream is = new TarArchiveInputStream(fis);
		TarArchiveEntry entry = is.getNextTarEntry();
		while (entry != null) {
			if (entry.isDirectory()) {
				continue;
			}
			if (Util.isImage(entry.getName())) {
				mEntradas.add(new TarEntry(entry, Util.toByteArray(is)));
			}
			entry = is.getNextTarEntry();
		}

		Collections.sort(mEntradas, new NaturalOrderComparator<TarEntry>() {
			@Override
			public String stringValue(TarEntry o) {
				return o.entry.getName();
			}
		}.thenComparing(new NaturalOrderComparator<TarEntry>() {
			@Override
			public String stringValue(TarEntry o) {
				return o.entry.getName();
			}
		}));
	}

	@Override
	public int getSize() {
		return mEntradas.size();
	}

	@Override
	public InputStream getPagina(int num) throws IOException {
		return new ByteArrayInputStream(mEntradas.get(num).bytes);
	}

	@Override
	public String getTipo() {
		return "tar";
	}

	@Override
	public void destroir() throws IOException {

	}

	@Override
	public List<String> getLegenda() {
		List<String> legendas = new ArrayList<String>();
		mLegendas.forEach((it) -> {
			InputStream sub = new ByteArrayInputStream(it.bytes);
			BufferedReader reader;
			try {
				reader = new BufferedReader(new InputStreamReader(sub, "UTF-8"));
				StringBuilder content = new StringBuilder();

				var line = reader.readLine();
				while (line != null) {
					content.append(line);
					line = reader.readLine();
				}

				legendas.add(content.toString());
			} catch (IOException e) {
				e.printStackTrace();
			}
		});
		return legendas;
	}

	@Override
	public Map<String, Integer> getLegendaNomes() {
		Map<String, Integer> arquivos = new HashMap<String, Integer>();

		for (var i = 0; i < mLegendas.size(); i++) {
			String path = Util.getNome(getName(mLegendas.get(i)));
			if (!path.isEmpty() && !arquivos.containsKey(path))
				arquivos.put(path, i);
		}

		return arquivos;
	}

	private String getName(TarEntry obj) {
		return obj.entry.getName();
	}

	@Override
	public String getPaginaPasta(Integer num) {
		if (mEntradas.size() < num)
            return null;
        return getName(mEntradas.get(num));
	}

	@Override
	public Map<String, Integer> getPastas() {
		Map<String, Integer> pastas = new HashMap<String, Integer>();

		for (var i = 0; i < mEntradas.size(); i++) {
			String path = Util.getPasta(getName(mEntradas.get(i)));
			if (!path.isEmpty() && !pastas.containsKey(path))
				pastas.put(path, i);
		}

        return pastas;
	}

}

File: ProcessaTexto/src/main/kotlin/br/com/fenix/processatexto/fileparse/TarParse.kt
package br.com.fenix.processatexto.fileparse

import br.com.fenix.processatexto.util.NaturalOrderComparator
import br.com.fenix.processatexto.util.Utils
import org.apache.commons.compress.archivers.tar.TarArchiveEntry
import org.apache.commons.compress.archivers.tar.TarArchiveInputStream
import org.slf4j.LoggerFactory
import java.io.*
import java.nio.charset.StandardCharsets
import java.util.*
import java.util.function.Consumer

class TarParse : Parse {

    private val LOGGER = LoggerFactory.getLogger(TarParse::class.java)

    private var mEntradas: MutableList<TarEntry> = ArrayList()
    private lateinit var mLegendas: List<TarEntry>

    private class TarEntry(val entry: TarArchiveEntry, val bytes: ByteArray)

    @Throws(IOException::class)
    override fun parse(file: File) {
        val fis = BufferedInputStream(FileInputStream(file))
        val `is` = TarArchiveInputStream(fis)
        var entry = `is`.nextTarEntry
        while (entry != null) {
            if (entry.isDirectory)
                continue

            if (Utils.isImage(entry.name))
                mEntradas.add(TarEntry(entry, Utils.toByteArray(`is`)))

            entry = `is`.nextTarEntry
        }
        Collections.sort(mEntradas, object : NaturalOrderComparator<TarEntry>() {
            override fun stringValue(o: TarEntry): String = o.entry.name
        }.thenComparing(object : NaturalOrderComparator<TarEntry>() {
            override fun stringValue(o: TarEntry): String = o.entry.name
        }))
    }

    override fun getSize(): Int = mEntradas.size

    @Throws(IOException::class)
    override fun getPagina(num: Int): InputStream = ByteArrayInputStream(mEntradas[num].bytes)

    override fun getTipo(): String = "tar"

    @Throws(IOException::class)
    override fun destroir() {
    }

    override fun getLegenda(): List<String> {
        val legendas: MutableList<String> = ArrayList()
        mLegendas.forEach(Consumer { it: TarEntry ->
            val sub: InputStream = ByteArrayInputStream(it.bytes)
            val reader: BufferedReader
            try {
                reader = BufferedReader(InputStreamReader(sub, StandardCharsets.UTF_8))
                val content = StringBuilder()
                var line = reader.readLine()
                while (line != null) {
                    content.append(line)
                    line = reader.readLine()
                }
                legendas.add(content.toString())
            } catch (e: IOException) {
                LOGGER.error(e.message, e)
            }
        })
        return legendas
    }

    override fun getLegendaNomes(): Map<String, Int> {
        val arquivos: MutableMap<String, Int> = HashMap()
        for (i in mLegendas.indices) {
            val path: String = Utils.getNome(getName(mLegendas[i]))
            if (path.isNotEmpty() && !arquivos.containsKey(path)) arquivos[path] = i
        }
        return arquivos
    }

    private fun getName(obj: TarEntry): String = obj.entry.name

    override fun getPaginaPasta(num: Int): String {
        return if (mEntradas.size < num) "" else getName(mEntradas[num])
    }

    override fun getPastas(): Map<String, Int> {
        val pastas: MutableMap<String, Int> = HashMap()
        for (i in mEntradas.indices) {
            val path: String = Utils.getPasta(getName(mEntradas[i]))
            if (path.isNotEmpty() && !pastas.containsKey(path)) pastas[path] = i
        }
        return pastas
    }
}